uitools/toolkitcpp/cpp/Esri/ArcGISRuntime/Toolkit/ArcGISAuthenticationController.cpp

```diff
@@ -0,0 +1,263 @@
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ******************************************************************************/
+#include "ArcGISAuthenticationController.h"
+
+// Qt headers
+#include <QFuture>
+#include <QPointer>
+#include <QUrl>
+
+// Maps SDK headers
+#include <ArcGISAuthenticationChallenge.h>
+#include <ArcGISAuthenticationManager.h>
+#include <ArcGISRuntimeEnvironment.h>
+#include <Error.h>
+#include <ErrorException.h>
+#include <OAuthUserConfiguration.h>
+#include <OAuthUserCredential.h>
+#include <OAuthUserLoginPrompt.h>
+#include <TokenCredential.h>
+
+using namespace Esri::ArcGISRuntime;
+
+namespace Esri::ArcGISRuntime::Toolkit {
+
+/*!
+ \class Toolkit::ArcGISAuthenticationController
+ \inmodule EsriArcGISRuntimeToolkit
+ \ingroup ArcGISQtToolkitUiCppControllers
+ \brief In MVC architecture, this is the controller for the corresponding
+        AuthenticationView.
+ */
+
+/*!
+  \brief Constructor
+  \list
+    \li \a parent Owning QObject.
+  \endlist
+ */
+ArcGISAuthenticationController::ArcGISAuthenticationController(QObject* parent) :
+  ArcGISAuthenticationChallengeHandler(parent)
+{
+  if (!canBeUsed_())
+    return;
+
+  // Set the authentication manager to use this controller
+  ArcGISRuntimeEnvironment::instance()->arcGISAuthenticationManager()->setArcGISAuthenticationChallengeHandler(this);
+
+  // listen for OAuth prompts
+  connect(ArcGISRuntimeEnvironment::instance()->arcGISAuthenticationManager(), &ArcGISAuthenticationManager::oAuthUserLoginPromptIssued, this,
+          [this](OAuthUserLoginPrompt* currentOAuthUserLoginPrompt)
+  {
+    m_currentOAuthUserLoginPrompt = currentOAuthUserLoginPrompt;
+    emit authorizeUrlChanged();
+    emit preferPrivateWebBrowserSessionChanged();
+    emit redirectUrlChanged();
+    emit displayOAuthSignInView();
+  });
+}
+
+/*!
+  \brief Destructor.
+ */
+ArcGISAuthenticationController::~ArcGISAuthenticationController() = default;
+
+ArcGISAuthenticationController* ArcGISAuthenticationController::create(QQmlEngine* qmlEngine, QJSEngine* /*jsEngine*/)
+{
+  static QPointer<ArcGISAuthenticationController> instance = new ArcGISAuthenticationController(qmlEngine);
+
+  if (!instance)
+    instance = new ArcGISAuthenticationController(qmlEngine);
+
+  return instance;
+}
+
+ArcGISAuthenticationController* ArcGISAuthenticationController::instance()
+{
+  return create(nullptr, nullptr);
+}
+
+void ArcGISAuthenticationController::handleArcGISAuthenticationChallenge(ArcGISAuthenticationChallenge* challenge)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
```

Comment:
Why a mutex for `handleArcGISAuthenticationChallenge`? Is it required by this specific implementation in toolkit or this should be done all the time when implementing `handleArcGISAuthenticationChallenge`? Is it documented somewhere?

Reply:
It is because users can set `m_userConfigurations` (indirectly via setter). It is highly unlikely both could happen at the same time, but I wanted the code to be safe and prevent any data races.